### PR TITLE
[faiss v1.7.1] Added high-level IVFFlatIndex impl

### PR DESCRIPTION
**Pull request by ava57r**
bindings.rs is from not released API.

### CODE REVIEW

**Comment by ava57r** [comment]
#13

File: src/index/ivf_flat.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            let mut inner = ptr::null_mut();
+            faiss_try(faiss_IndexIVFFlat_new_with_metric(
+                &mut inner,
+                quantizer.inner_ptr(),
+                d as usize,
+                nlist as usize,
+                metric,
+            ))?;
+
+            mem::forget(quantizer); // own_fields default == true
```
Comment (ava57r) [inline src/index/ivf_flat.rs:46]: leak is here
IVF inherited from `Level1Quantizer`
https://github.com/facebookresearch/faiss/blob/master/faiss/IndexIVF.h#L95

`own_fields` default == false https://github.com/facebookresearch/faiss/blob/master/faiss/IndexIVF.cpp#L40

**Comment by ava57r** [comment]
https://github.com/facebookresearch/faiss/pull/1827

**Comment by ava57r** [comment]
PR https://github.com/facebookresearch/faiss/pull/1787 merged

**Comment by ava57r** [comment]
PR https://github.com/facebookresearch/faiss/pull/1764 merged

**Comment by ava57r** [comment]
I finished.
You can review it

File: src/index/ivf_flat.rs [hunk 2]
```diff
@@ -57,20 +57,18 @@ impl IVFFlatIndexImpl {
                 nlist as usize,
                 metric,
             ))?;
-            let own_fields_ = if own_fields { 1 } else { 0 };
-            faiss_IndexIVFFlat_set_own_fields(inner, own_fields_);
-
+            faiss_IndexIVFFlat_set_own_fields(inner, own_fields as i32);
```
Comment (Enet4) [inline src/index/ivf_flat.rs:60]: IIRC the effective argument type of `own_fields` is `c_int`, so we might as well follow that. Please try either `c_int::from(own_fields)` or `own_fields as c_int` if the former does not work.

File: src/index/ivf_flat.rs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        }
+    }
+}
+
+/**
+ * = 0: use the quantizer as index in a kmeans training
+ * = 1: just pass on the training set to the train() of the quantizer
+ * = 2: kmeans training on a flat index + add the centroids to the quantizer
+ */
+#[derive(Debug, Clone)]
```
Comment (Enet4) [inline src/index/ivf_flat.rs:127]: We can derive a few more traits on this one, as per the [C-COMMON-TRAITS](https://rust-lang.github.io/api-guidelines/interoperability.html#c-common-traits) guideline.
```suggestion
#[derive(Debug, Copy, Clone, Eq, Hash, PartialEq)]
```

**Review by Enet4 — COMMENTED** [review]
We will be in great condition to release a new version of `faiss-rs` once these are merged. :)

I will try to review them all within the next few weeks.

File: src/index/ivf_flat.rs (hunk 2 above)
Comment (ava57r) [reply src/index/ivf_flat.rs:60]: OK

**Review by Enet4 — APPROVED** [review]
Looks OK, thanks!

**Pull request merged**